### ISSUE TITLE
Issue in the colmap_loader.py

### ISSUE THREAD

**Issue by wenxuanou** [graphdeco-inria/gaussian-splatting#432]
In [line 65 of scene/colmap_loader.py](https://github.com/graphdeco-inria/gaussian-splatting/blob/2eee0e26d2d5fd00ec462df47752223952f6bf4e/scene/colmap_loader.py#L65), there is a if statement multiplying -1 to qvec when q[0] is negative, which makes the output of ``qvec2rotmat(rotmat2qvec(quat))`` unmatch to the input ``quat``.

What is the reason for multiplying a -1 when computing the quaternion? 

Below is a script I wrote comparing the quaternion computation in Gaussian-splatting and the pytorch implementation.
```
import numpy as np
import torch
import torch.nn.functional as F

# Gaussian-splatting implementation
def qvec2rotmat(qvec):
    return np.array([
        [1 - 2 * qvec[2]**2 - 2 * qvec[3]**2,
         2 * qvec[1] * qvec[2] - 2 * qvec[0] * qvec[3],
         2 * qvec[3] * qvec[1] + 2 * qvec[0] * qvec[2]],
        [2 * qvec[1] * qvec[2] + 2 * qvec[0] * qvec[3],
         1 - 2 * qvec[1]**2 - 2 * qvec[3]**2,
         2 * qvec[2] * qvec[3] - 2 * qvec[0] * qvec[1]],
        [2 * qvec[3] * qvec[1] - 2 * qvec[0] * qvec[2],
         2 * qvec[2] * qvec[3] + 2 * qvec[0] * qvec[1],
         1 - 2 * qvec[1]**2 - 2 * qvec[2]**2]])

def rotmat2qvec(R):
    Rxx, Ryx, Rzx, Rxy, Ryy, Rzy, Rxz, Ryz, Rzz = R.flat
    K = np.array([
        [Rxx - Ryy - Rzz, 0, 0, 0],
        [Ryx + Rxy, Ryy - Rxx - Rzz, 0, 0],
        [Rzx + Rxz, Rzy + Ryz, Rzz - Rxx - Ryy, 0],
        [Ryz - Rzy, Rzx - Rxz, Rxy - Ryx, Rxx + Ryy + Rzz]]) / 3.0
    eigvals, eigvecs = np.linalg.eigh(K)
    qvec = eigvecs[[3, 0, 1, 2], np.argmax(eigvals)]
    #if qvec[0] < 0:
    #    qvec *= -1
    return qvec

# Pytorch implementation (adjusted for numpy)
def _sqrt_positive_part(x):
    """
    Returns torch.sqrt(torch.max(0, x))
    but with a zero subgradient where x is 0.
    """
    ret = torch.zeros_like(x)
    positive_mask = x > 0
    ret[positive_mask] = torch.sqrt(x[positive_mask])
    return ret
    
def matrix_to_quaternion(matrix):
    """
    Convert rotations given as rotation matrices to quaternions.

    Args:
        matrix: Rotation matrices as tensor of shape (..., 3, 3).

    Returns:
        quaternions with real part first, as tensor of shape (..., 4).
    """
    matrix = torch.from_numpy(matrix)
    with torch.no_grad():
        if matrix.size(-1) != 3 or matrix.size(-2) != 3:
            raise ValueError(f"Invalid rotation matrix shape {matrix.shape}.")

        batch_dim = matrix.shape[:-2]
        m00, m01, m02, m10, m11, m12, m20, m21, m22 = torch.unbind(
            matrix.reshape(batch_dim + (9,)), dim=-1
        )

        q_abs = _sqrt_positive_part(
            torch.stack(
                [
                    1.0 + m00 + m11 + m22,
                    1.0 + m00 - m11 - m22,
                    1.0 - m00 + m11 - m22,
                    1.0 - m00 - m11 + m22,
                ],
                dim=-1,
            )
        )

        # we produce the desired quaternion multiplied by each of r, i, j, k
        quat_by_rijk = torch.stack(
            [
                torch.stack([q_abs[..., 0] ** 2, m21 - m12, m02 - m20, m10 - m01], dim=-1),
                torch.stack([m21 - m12, q_abs[..., 1] ** 2, m10 + m01, m02 + m20], dim=-1),
                torch.stack([m02 - m20, m10 + m01, q_abs[..., 2] ** 2, m12 + m21], dim=-1),
                torch.stack([m10 - m01, m20 + m02, m21 + m12, q_abs[..., 3] ** 2], dim=-1),
            ],
            dim=-2,
        )

        # We floor here at 0.1 but the exact level is not important; if q_abs is small,
        # the candidate won't be picked.
        flr = torch.tensor(0.1).to(dtype=q_abs.dtype, device=q_abs.device)
        quat_candidates = quat_by_rijk / (2.0 * q_abs[..., None].max(flr))

        # if not for numerical problems, quat_candidates[i] should be same (up to a sign),
        # forall i; we pick the best-conditioned one (with the largest denominator)

        quat_candidates = quat_candidates[
            F.one_hot(q_abs.argmax(dim=-1), num_classes=4) > 0.5, :  # pyre-ignore[16]
        ].reshape(batch_dim + (4,))
    return quat_candidates.cpu().detach().numpy()


if __name__ == "__main__":
    quat = np.array([-0.24186184, -0.26458945, -0.64748156, 0.67250494])

    rot = qvec2rotmat(quat)

    quat2 = rotmat2qvec(rot)
    quat3 = matrix_to_quaternion(rot)

    print("original quat: ", quat)
    print("rot: ", rot)

    print("output quat from 3dgs: ", quat2)  # this gives [ 0.24186184  0.26458945  0.64748156 -0.67250494]
    print("output quat from torch: ", quat3)  # this gives [-0.24186184 -0.26458945 -0.64748156  0.67250494]
```

**Comment by yuedajiong** [comment]
My Code:

#1:
    from scipy.spatial.transform import Rotation
    Q = Rotation.from_matrix(R).as_quat()  #canonical=True

#2:
def q_to_R(self, q):  # [...,4]  # https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion
    qa, qb, qc, qd = q.unbind(dim=-1)
    R = torch.stack(
        [torch.stack([1 - 2 * (qc ** 2 + qd ** 2), 2 * (qb * qc - qa * qd), 2 * (qa * qc + qb * qd)], dim=-1),
         torch.stack([2 * (qb * qc + qa * qd), 1 - 2 * (qb ** 2 + qd ** 2), 2 * (qc * qd - qa * qb)], dim=-1),
         torch.stack([2 * (qb * qd - qa * qc), 2 * (qa * qb + qc * qd), 1 - 2 * (qb ** 2 + qc ** 2)], dim=-1)],
        dim=-2)
    return R

def R_to_q(self, R, eps=1e-6):  # [...,3,3]  # https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion
    row0, row1, row2 = R.unbind(dim=-2)
    R00, R01, R02 = row0.unbind(dim=-1)
    R10, R11, R12 = row1.unbind(dim=-1)
    R20, R21, R22 = row2.unbind(dim=-1)
    t = R[..., 0, 0] + R[..., 1, 1] + R[..., 2, 2]
    r = (1 + t + eps).sqrt()
    qa = 0.5 * r
    qb = (R21 - R12).sign() * 0.5 * (1 + R00 - R11 - R22 + eps).sqrt()
    qc = (R02 - R20).sign() * 0.5 * (1 - R00 + R11 - R22 + eps).sqrt()
    qd = (R10 - R01).sign() * 0.5 * (1 - R00 - R11 + R22 + eps).sqrt()
    q = torch.stack([qa, qb, qc, qd], dim=-1)
    return q



[superi_coord.py.txt](https://github.com/graphdeco-inria/gaussian-splatting/files/13266847/superi_coord.py.txt)

**Comment by kwea123** [comment]
I didn't read the whole code, but quaternion is [invariant to multiplying -1](https://math.stackexchange.com/questions/2016282/negative-quaternion) i.e. `q` and `-q` represent the same rotation

**Comment by wenxuanou** [comment]
> I didn't read the whole code, but quaternion is [invariant to multiplying -1](https://math.stackexchange.com/questions/2016282/negative-quaternion) i.e. `q` and `-q` represent the same rotation

I added ``quaternion_to_axis_angle()`` to see the effect of -1 to the angles. It turns out that the resulting axis angle for ``q`` and ``-q`` are not the same. 

Here is my updated script:
```
import numpy as np
import torch
import torch.nn.functional as F

# Gaussian-splatting implementation
def qvec2rotmat(qvec):
    return np.array([
        [1 - 2 * qvec[2]**2 - 2 * qvec[3]**2,
         2 * qvec[1] * qvec[2] - 2 * qvec[0] * qvec[3],
         2 * qvec[3] * qvec[1] + 2 * qvec[0] * qvec[2]],
        [2 * qvec[1] * qvec[2] + 2 * qvec[0] * qvec[3],
         1 - 2 * qvec[1]**2 - 2 * qvec[3]**2,
         2 * qvec[2] * qvec[3] - 2 * qvec[0] * qvec[1]],
        [2 * qvec[3] * qvec[1] - 2 * qvec[0] * qvec[2],
         2 * qvec[2] * qvec[3] + 2 * qvec[0] * qvec[1],
         1 - 2 * qvec[1]**2 - 2 * qvec[2]**2]])
def rotmat2qvec(R):
    Rxx, Ryx, Rzx, Rxy, Ryy, Rzy, Rxz, Ryz, Rzz = R.flat
    K = np.array([
        [Rxx - Ryy - Rzz, 0, 0, 0],
        [Ryx + Rxy, Ryy - Rxx - Rzz, 0, 0],
        [Rzx + Rxz, Rzy + Ryz, Rzz - Rxx - Ryy, 0],
        [Ryz - Rzy, Rzx - Rxz, Rxy - Ryx, Rxx + Ryy + Rzz]]) / 3.0
    eigvals, eigvecs = np.linalg.eigh(K)
    qvec = eigvecs[[3, 0, 1, 2], np.argmax(eigvals)]
    if qvec[0] < 0:
        qvec *= -1
    return qvec

####################################################
# Pytorch implementation (adjusted for numpy)
def _sqrt_positive_part(x):
    """
    Returns torch.sqrt(torch.max(0, x))
    but with a zero subgradient where x is 0.
    """
    ret = torch.zeros_like(x)
    positive_mask = x > 0
    ret[positive_mask] = torch.sqrt(x[positive_mask])
    return ret
def matrix_to_quaternion(matrix):
    """
    Convert rotations given as rotation matrices to quaternions.
    Args:
        matrix: Rotation matrices as tensor of shape (..., 3, 3).
    Returns:
        quaternions with real part first, as tensor of shape (..., 4).
    """
    matrix = torch.from_numpy(matrix)
    with torch.no_grad():
        if matrix.size(-1) != 3 or matrix.size(-2) != 3:
            raise ValueError(f"Invalid rotation matrix shape {matrix.shape}.")
        batch_dim = matrix.shape[:-2]
        m00, m01, m02, m10, m11, m12, m20, m21, m22 = torch.unbind(matrix.reshape(batch_dim + (9,)), dim=-1)
        q_abs = _sqrt_positive_part(
            torch.stack([
                    1.0 + m00 + m11 + m22,
                    1.0 + m00 - m11 - m22,
                    1.0 - m00 + m11 - m22,
                    1.0 - m00 - m11 + m22,],
                dim=-1,
            ) )

        # we produce the desired quaternion multiplied by each of r, i, j, k
        quat_by_rijk = torch.stack([
                torch.stack([q_abs[..., 0] ** 2, m21 - m12, m02 - m20, m10 - m01], dim=-1),
                torch.stack([m21 - m12, q_abs[..., 1] ** 2, m10 + m01, m02 + m20], dim=-1),
                torch.stack([m02 - m20, m10 + m01, q_abs[..., 2] ** 2, m12 + m21], dim=-1),
                torch.stack([m10 - m01, m20 + m02, m21 + m12, q_abs[..., 3] ** 2], dim=-1),
        ],dim=-2,)

        # We floor here at 0.1 but the exact level is not important; if q_abs is small,
        # the candidate won't be picked.
        flr = torch.tensor(0.1).to(dtype=q_abs.dtype, device=q_abs.device)
        quat_candidates = quat_by_rijk / (2.0 * q_abs[..., None].max(flr))

        # if not for numerical problems, quat_candidates[i] should be same (up to a sign),
        # forall i; we pick the best-conditioned one (with the largest denominator)

        quat_candidates = quat_candidates[
            F.one_hot(q_abs.argmax(dim=-1), num_classes=4) > 0.5, :  # pyre-ignore[16]
        ].reshape(batch_dim + (4,))
    return quat_candidates.cpu().detach().numpy()

# Borrow from pytorch again
def quaternion_to_axis_angle(quaternions):
    """
    Convert rotations given as quaternions to axis/angle.
    Args:
        quaternions: quaternions with real part first,
            as tensor of shape (..., 4).
    Returns:
        Rotations given as a vector in axis angle form, as a tensor
            of shape (..., 3), where the magnitude is the angle
            turned anticlockwise in radians around the vector's
            direction.
    """
    quaternions = torch.from_numpy(quaternions)
    with torch.no_grad():
        norms = torch.norm(quaternions[..., 1:], p=2, dim=-1, keepdim=True)
        half_angles = torch.atan2(norms, quaternions[..., :1])
        angles = 2 * half_angles
        eps = 1e-6
        small_angles = angles.abs() < eps
        sin_half_angles_over_angles = torch.empty_like(angles)
        sin_half_angles_over_angles[~small_angles] = (
            torch.sin(half_angles[~small_angles]) / angles[~small_angles]
        )
        # for x small, sin(x/2) is about x/2 - (x/2)^3/6
        # so sin(x/2)/x is about 1/2 - (x*x)/48
        sin_half_angles_over_angles[small_angles] = (
            0.5 - (angles[small_angles] * angles[small_angles]) / 48
        )
        quaternions = quaternions[..., 1:] / sin_half_angles_over_angles
    return quaternions.cpu().detach().numpy()
    


if __name__ == "__main__":
    quat = np.array([-0.24186184, -0.26458945, -0.64748156, 0.67250494]) # this comes from a sample from my dataset

    rot = qvec2rotmat(quat)

    quat2 = rotmat2qvec(rot)
    quat3 = matrix_to_quaternion(rot)

    print("original quat: ", quat)
    print("rot: ", rot)

    print("output quat from 3dgs: ", quat2)  # this gives [ 0.24186184  0.26458945  0.64748156 -0.67250494]
    print("output quat from torch: ", quat3) # this gives [-0.24186184 -0.26458945 -0.64748156  0.67250494]

    angle = quaternion_to_axis_angle(quat2) * 180.0 / np.pi
    angle2 = quaternion_to_axis_angle(quat3) * 180.0 / np.pi

    print("axis angle from 3dgs quat (in degrees): ", angle)  # this gives [  41.45009756  101.43327267 -105.35338945]
    print("axis angle from torch quat (in degrees): ", angle2) # this gives [ -56.71660549 -138.79221639  144.1561535 ]  
    print("axis angle from torch quat + 360 degree: ", angle2 + 360.0) # this gives [303.28339451 221.20778361 504.1561535 ] 
    
```

The pytorch implementation can be found [here](https://pytorch3d.readthedocs.io/en/latest/_modules/pytorch3d/transforms/rotation_conversions.html).
